x-pack/plugins/uptime/server/lib/helper/format_es_buckets_for_histogram.ts

```diff
@@ -0,0 +1,44 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { UMESBucket, UMESHistogramBucket } from '../adapters/database';
+
+/**
+ * The charting library we're currently using requires histogram data points have an
+ * x and an x0 property, where x0 is the beginning of a data point and x provides
+ * the size of the point from the start. This function attempts to generalize the
+ * concept so any bucket that has a numeric value as its key can be put into this format.
+ *
+ * Additionally, histograms that stack horizontally instead of vertically need to have
+ * a y and a y0 value. We're not doing this currently but with some minor modification
+ * this function could provide formatting for those buckets as well.
+ * @param buckets The ES data to format.
+ */
+export function formatEsBucketsForHistogram<T extends UMESBucket>(
+  buckets: T[]
+): Array<T & UMESHistogramBucket> {
+  // wait for first bucket to fill up
+  if (buckets.length < 2) {
+    return [];
+  }
+  const TERMINAL_INDEX = buckets.length - 1;
+  const { key: terminalBucketTime } = buckets[TERMINAL_INDEX];
+  // drop the most recent bucket to avoid returning incomplete bucket
+  return buckets.slice(0, TERMINAL_INDEX).map((item, index, array) => {
+    const { key } = item;
+    const nextItem = array[index + 1];
+    const bucketSize = nextItem ? Math.abs(nextItem.key - key) : Math.abs(terminalBucketTime - key);
+
+    return Object.assign(
```

Comment:
TIL about Object.assign

Reply:
Fwiw you could also do this for the same effect.

return {
        x: key + bucketSize,
        x0: key,
        ...item,
};

Reply:
@mattapperson funny you mention that, it's _exactly_ what I had!
I ran `node ./scripts/type_check.js` in preparation for this PR and it blew up on that, said that spread assignments can only be used for objects. I guess maybe because I'm using the generic `T extends {...}` here?
Anyway, I'd prefer to use the code you just showed, but dunno how to please the linter.
It doesn't complain about the use of `Object.assign(...)`, which is basically a more verbose way of doing the same.

Reply:
Ok - so the reason this is happening is because I was running TS `3.2.2` when I wrote this code instead of my workspace version, `3.0.3`.

@mattapperson - given that this code works with `Object.assign`, is tested, and those tests are passing our CI, would you be ok with me tracking this in an issue and refactoring later when code like
```
return {
  x: key + bucketSize,
  x0: key,
  ...item,
};
```
will be allowed by our supported TS version?

It appears to me I'm trying to take advantage of a feature that was deemed to be a part of TS's design goals, but wasn't yet added to the version we're running today; so I'd be ok with a temporary workaround. WDYT?

Reply:
I'm going to track this [here](https://github.com/elastic/kibana/issues/30688), we will update the code when we're able.